# Exercise 2.A Fundamentals of pandas

In [1]:
import numpy as np
import pandas as pd

## Series

In [10]:
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}

In [11]:
print('aa:', aa)
print('ab:', ab)
print('ac:', ac)
print('ad:', ad)

aa: [1, 2, 3, 4]
ab: (5, 6, 7, 8)
ac: {9, 10, 11, 12}
ad: {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}


In [12]:
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(list(ac))  # wrap the set in list() — pandas needs an ordered collection
ser4 = pd.Series(ad)

In [13]:
print('ser1:')
print(ser1)
print('\nser2:')
print(ser2)
print('\nser3:')
print(ser3)
print('\nser4:')
print(ser4)

ser1:
0    1
1    2
2    3
3    4
dtype: int64

ser2:
0    5
1    6
2    7
3    8
dtype: int64

ser3:
0     9
1    10
2    11
3    12
dtype: int64

ser4:
first     13
second    14
third     15
fourth    16
dtype: int64


## DataFrames from Series

In [14]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [15]:
df1.insert(1, 'col2', ser2)
df1.insert(2, 'col3', ser3)
df1

,0,col2,col3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [ ]:
df1.insert(3, 'col4', ser4)
df1
# ser4's index was 'first', 'second', 'third', 'fourth'.
# df1's index is 0, 1, 2, 3.
# The indexes don't match, so pandas can't line ser4 up it fills the whole column with NaN

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [17]:
df1.loc[4] = [5, 9, 12, 0]
df1

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [19]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [20]:
df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [21]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [22]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        262 non-null    object
 1   description  262 non-null    object
 2   category     262 non-null    object
 3   language     262 non-null    object
 4   actor_name   262 non-null    object
dtypes: object(5)
memory usage: 10.4+ KB


A few things I can tell from df2.info():
- 262 rows, 5 columns
- All columns are  strings, no numeric data
- No nulls in any column, the data is complete
- The data is about movies and their actors, with category and language info

In [23]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


all columns are text, describe() shows count, unique, top, and freq instead of mean/std. Useful: language only has 1 unique value, and category tells us how many genres exist. Less useful: top/freq on description doesn't really help since most descriptions are unique.

In [24]:
df3 = pd.read_csv('northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [25]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


describe() shows the numeric stats for OrderID, UnitPrice, and Quantity. ProductName is text so it gets skipped here. The difference is that Northwind has actual numbers in it, while Sakila was all text. OrderID gets statistics calculated on it even though averaging an ID number is meaningless. The useful stats are on UnitPrice and Quantity, it tells us the price range and how much customers typically order

## Coding a DataFrame

In [26]:
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [27]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [28]:
df4.to_csv('test_data.csv', index=False)